Core: Fix incremental compute of partition stats with COW deletes

ajantha-bhat opened the pull request:
Fixes an edge case where deleted entires information is not carried to next snapshot. So, incremental stats compute was wrong for this copy on write case. Fix is to apply incremental stats compute snapshot by snapshot for its added manifest.

Also fallback when stats file exist for table but not for current snapshot chain, instead of throwing exception.  

Fixes: #13155

lirui-apache commented on core/src/test/java/org/apache/iceberg/PartitionStatsHandlerTestBase.java:
```diff
@@ -441,6 +441,52 @@ public void testPartitionStats() throws Exception {
+    assertThat(
+            PartitionStatsHandler.readPartitionStatsFile(
+                PartitionStatsHandler.schema(Partitioning.partitionType(testTable)),
+                Files.localInput(statisticsFileNew.path())))
```
BTW, I noted Files.localInput is also used in production code [PartitionStatsHandler::computeAndMergeStatsIncremental](https://github.com/apache/iceberg/blob/cab0decbb0e32bf314039e30807eb033c50665d5/core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java#L279) to load the previous stats file. Shouldn't we use table's FileIO to do that?

ajantha-bhat replied:
done

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java:
```diff
@@ -276,7 +275,8 @@ private static Collection<PartitionStats> computeAndMergeStatsIncremental(
     try (CloseableIterable<PartitionStats> oldStats =
-        readPartitionStatsFile(schema(partitionType), Files.localInput(previousStatsFile.path()))) {
+        readPartitionStatsFile(
+            schema(partitionType), table.io().newInputFile(previousStatsFile.path()))) {
```
Just using the FileIO as pointed out in the PR. 
Unrelated to this bug fix. But related to this feature.

ajantha-bhat commented on core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java:
```diff
@@ -336,16 +336,29 @@ private static PartitionMap<PartitionStats> computeStatsDiff(
 
   private static PartitionMap<PartitionStats> computeStats(
-      Table table, Snapshot snapshot, Predicate<ManifestFile> predicate, boolean incremental) {
+      Table table, Snapshot snapshot, Set<Long> snapshotIdsRange) {
```
Done. 

I had considered it earlier as well, but decided to leave it as-is since I felt I might be overthinking the refactoring of this private method. But it makes sense now as it is very clean now.

ajantha-bhat commented:
@lirui-apache: Could you please take another look at the fix, verify and  approve the PR if it looks good for you?

lirui-apache commented on core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java:
```diff
@@ -336,17 +337,23 @@ private static PartitionMap<PartitionStats> computeStatsDiff(
+    // So, for incremental computation, gather the manifests added by each snapshot
+    // instead of relying solely on those from the latest snapshot.
+    List<ManifestFile> manifests =
+        snapshotIdsRange.stream()
```
nit: How about just call [SnapshotUtil::ancestorsBetween](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L208) and iterate through the ancestors?

ajantha-bhat replied:
Updated.

lirui-apache reviewed (commented):
Thanks for the fix @ajantha-bhat . I only have a minor comment. But I have another question related to the feature and may be worth another PR. In [PartitionStatsHandler::latestStatsFile](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java#L329), we throw exception if there is any previous stats file but not reachable via the parent snapshot pointers. I think this means we don't allow gaps in snapshot history, which can be common when using tags. Won't it be more user friendly to just fall back to a full compute in that case?

ajantha-bhat commented:
> But I have another question related to the feature and may be worth another PR. In [PartitionStatsHandler::latestStatsFile](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/PartitionStatsHandler.java#L329), we throw exception if there is any previous stats file but not reachable via the parent snapshot pointers. I think this means we don't allow gaps in snapshot history, which can be common when using tags. Won't it be more user friendly to just fall back to a full compute in that case?

Great. I didn't think of this case before. Hence, added that exception before. I made it to fallback now and added the test.

lirui-apache reviewed (approved):
Thanks for updating. LGTM
